presto-hive-common/src/main/java/com/facebook/presto/hive/zorder/ZOrderCurveSearcher.java

```diff
@@ -200,4 +161,47 @@ private boolean outOfRange(List<Integer> startValues, List<Integer> endValues, L
         }
         return false;
     }
+
+    private static List<ZAddressRange<Long>> combineAddressRanges(List<ZAddressRange<Long>> addressRanges)
+    {
+        if (addressRanges.isEmpty()) {
+            return addressRanges;
+        }
+
+        List<ZAddressRange<Long>> sortedAddressRanges = new ArrayList<>(addressRanges);
+        sortedAddressRanges.sort((range1, range2) -> {
+            if (range1.equals(range2)) {
+                return 0;
+            }
+            if (range1.getMinimumAddress().equals(range2.getMinimumAddress())) {
+                return (range1.getMaximumAddress() > range2.getMaximumAddress()) ? 1 : -1;
+            }
+            return (range1.getMinimumAddress() > range2.getMinimumAddress()) ? 1 : -1;
+        });
+
+        List<ZAddressRange<Long>> combinedAddressRanges = new ArrayList<>();
+        combinedAddressRanges.add(addressRanges.get(0));
+
+        for (int index = 1; index < addressRanges.size(); index++) {
+            ZAddressRange<Long> previousAddressRange = combinedAddressRanges.get(combinedAddressRanges.size() - 1);
+            ZAddressRange<Long> currentAddressRange = addressRanges.get(index);
+
+            if (isOverlapping(previousAddressRange, currentAddressRange)) {
+                combinedAddressRanges.set(combinedAddressRanges.size() - 1, new ZAddressRange<>(previousAddressRange.getMinimumAddress(), currentAddressRange.getMaximumAddress()));
+            }
+            else {
+                combinedAddressRanges.add(currentAddressRange);
+            }
+        }
+
+        return combinedAddressRanges;
+    }
+
+    private static boolean isOverlapping(ZAddressRange<Long> previousAddressRange, ZAddressRange<Long> currentAddressRange)
+    {
+        return (previousAddressRange.getMinimumAddress().equals(currentAddressRange.getMinimumAddress()))
+                || (previousAddressRange.getMaximumAddress().equals(currentAddressRange.getMaximumAddress()))
+                || (previousAddressRange.getMaximumAddress() + 1 == currentAddressRange.getMinimumAddress())
+                || (previousAddressRange.getMaximumAddress() > currentAddressRange.getMinimumAddress());
+    }
```

Comment:
When the following condition is satisfied, these two range should be merged. Do you think so?

`previousAddressRange.getMaximumAddress() + 1 >= currentAddressRange.getMinimumAddress()`